python/ray/dataframe/dataframe.py

```diff
@@ -2882,35 +2881,72 @@ def quantile(self, q=0.5, axis=0, numeric_only=True,
+            for t in self.dtypes:
+                if t == np.dtype('O'):
+                    e_msg = "Cannot call quantile on columns of type " + str(t)
+                    raise TypeError(e_msg)
+        else:
+            if all([dtype == np.dtype('O') or
+                    np.issubdtype(dtype, np.datetime64)
+                    for dtype in self.dtypes]):
+                raise ValueError("need at least one array to concatenate")
+
+        # check that all qs are between 0 and 1
+        pd.DataFrame()._check_percentile(q)
 
         if isinstance(q, (pd.Series, np.ndarray, pd.Index, list)):
-            # In the case of a list, we build it one at a time.
-            # TODO Revisit for performance
-            quantiles = []
-            for q_i in q:
-                def remote_func(df):
-                    return quantile_helper(df, q=q_i, axis=axis,
-                                           numeric_only=numeric_only,
-                                           interpolation=interpolation)
-
-                result = self._arithmetic_helper(remote_func, axis)
-                result.name = q_i
-                quantiles.append(result)
-
-            return pd.concat(quantiles, axis=1).T
+
+            def quantile_helper(df):
+                # This if call prevents ValueErrors with object only partitions
+                if (numeric_only and
+                        all([(dtype == np.dtype('O') or
+                              np.issubdtype(dtype, np.datetime64))
+                            for dtype in df.dtypes])):
+                    return pd.DataFrame()
+                else:
+                    return df.quantile(q=q, axis=axis,
+                                       numeric_only=numeric_only,
+                                       interpolation=interpolation)
+
+            if axis == 0 or axis == 'index':
+                result = _map_partitions(quantile_helper,
+                                         self._col_partitions)
+                q_index = pd.Float64Index(q)
+
+                # select only correct dtype columns
+                new_cols = self.dtypes[self.dtypes.apply(
+                                       lambda x: is_numeric_dtype(x))].index
+                return DataFrame(col_partitions=result,
+                                 index=q_index,
+                                 columns=new_cols)
+
+            if axis == 1 or axis == 'columns':
+                result = _map_partitions(quantile_helper,
+                                         self._row_partitions)
+                q_index = pd.Float64Index(q)
+                return DataFrame(col_partitions=result,
+                                 index=q_index,
+                                 columns=self.index)
+
+        # single q means we can use arithmetic helper
         else:
-            def remote_func(df):
-                return quantile_helper(df, q=q, axis=axis,
+
+            def quantile_helper(df):
```

Comment:
Why is this function defined twice?

Reply:
One returns a Series the other returns a DataFrame